### PR TITLE
fire embedding and clustering at end of pipeline

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-862

#### Link to staging deployment URL 

In UI repo : https://github.com/biomage-ltd/ui/pull/230

#### Links to any Pull Requests related to this

UI - https://github.com/biomage-ltd/ui/pull/230
API - https://github.com/biomage-ltd/api/pull/99
Pipeline - https://github.com/biomage-ltd/pipeline/pull/53

#### Anything else the reviewers should know about the changes here

- Fire worker embedding and clustering worker actions
- Created new `hookRunner` function to register and run hooks

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [X] Unit tests written
- [X] Tested locally with Inframock
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/api/route-services/pipeline-response.js [hunk 1]
```diff
@@ -78,9 +80,54 @@ const pipelineResponse = async (io, message) => {
...
+      uuid: uuidv4(),
+      body: {
+        name: 'GetEmbedding',
+        type: output.config.embeddingSettings.method,
+        config: output.config.embeddingSettings.methodSettings[
+          output.config.embeddingSettings.method
+        ],
+      },
+      PipelineRunETag: statusRes.pipeline.startDate,
+    };
```
Comment (ivababukova) [inline src/api/route-services/pipeline-response.js]: I am slightly concerned that this piece of code as it stands is very coupled with the schema of the compute embedding task and if we wanted to change it in the future, which makes it very fragile and easy to break. Is there a way to abstract away the specific parameter setup and have it more generic here?

File: src/api/route-services/pipeline-response.js [hunk 2]
```diff
@@ -78,9 +80,54 @@ const pipelineResponse = async (io, message) => {
...
+        type: output.config.clusteringSettings.method,
+        cellSetKey: 'louvain',
+        config: {
+          resolution: output.config.clusteringSettings.methodSettings[
+            output.config.clusteringSettings.method
+          ],
+        },
+      },
+      PipelineRunETag: statusRes.pipeline.startDate,
+    };
```
Comment (ivababukova) [inline src/api/route-services/pipeline-response.js]: same here - this piece of code is easy to break -- all it takes is to do a tiny change in the task schema.

File: src/api/route-services/pipeline-response.js [hunk 3]
```diff
@@ -78,9 +80,54 @@ const pipelineResponse = async (io, message) => {
     ]);
   }
 
-
   const statusRes = await pipelineStatus(experimentId);
 
+  if (taskName === 'configureEmbedding') {
+    const now = new Date();
```
Comment (xverges) [inline src/api/route-services/pipeline-response.js]: Can we extract this to its own function? And, to ease @ivababukova's concerns, could we explicitly validate the work requests here?

Comment (aerlaut) [reply src/api/route-services/pipeline-response.js]: Extracted into `hookRunner` and associated functions under `src/utils/hooks`

File: src/api/route-services/pipeline-response.js (hunk 2 above)
Comment (aerlaut) [reply src/api/route-services/pipeline-response.js]: This is refactored into `hookRunner` and associated functions under `src/utils/hooks`

File: src/api/route-services/pipeline-response.js (hunk 1 above)
Comment (aerlaut) [reply src/api/route-services/pipeline-response.js]: This is refactored into `hookRunner` and associated functions under `src/utils/hooks`

**Review by ivababukova — APPROVED** [review]

**Pull request merged**